tests/migrate_signals/test_operation_signals.py

```diff
@@ -0,0 +1,152 @@
+from unittest.mock import patch
+
+from django.db import connection
+from django.db.migrations import Migration
+from django.db.migrations.executor import MigrationExecutor
+from django.db.migrations.operations import RunPython
+from django.db.migrations.recorder import MigrationRecorder
+from django.db.models import signals
+from django.test import TransactionTestCase, override_settings
+
+from .operations import Operation0, Operation1, Operation2, Operation3
+
+
+class Receiver:
+    def __init__(self, operation):
+        self.sender = operation
+        signals.post_operation.connect(self, sender=operation)
+
+
+class InjectOperation1Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation1()]
+
+
+class InjectOperation2Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation2()]
+
+
+class InjectOperation1And2Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation1(), Operation2()]
+
+
+class InjectOperation3Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation3()]
+
+
+@override_settings(
+    MIGRATION_MODULES={"migrate_signals": "migrate_signals.noop_migrations"}
+)
+class PostOperationSignalTests(TransactionTestCase):
+
+    available_apps = ["migrate_signals"]
+
+    def setUp(self):
+        self.receivers = []
+
+    def tearDown(self):
+        # Disconnect all the receivers
+        for receiver in self.receivers:
+            signals.post_operation.disconnect(receiver, sender=receiver.sender)
+
+        # Reset applied-migrations state.
+        recorder = MigrationRecorder(connection)
+        recorder.migration_qs.filter(app='migrate_signals').delete()
+
+    def _register(self, inject_operation):
+        return self.receivers.append(inject_operation)
+
+    def test_apply_infinite_recursion(self):
+        self._register(InjectOperation1Receiver(Operation0))
+        self._register(InjectOperation1Receiver(Operation1))
```

Comment:
Unless I'm missing something a single `InjectOperation0Receiver(Operation0)` (or `Receiver(Operation0, [Operation0])`) should do here. Could be worth testing both cases, direct or cyclical recursion like you did here.

Reply:
Yes you are right, it's because I was using the `CreateModel` from `custom_migrations` before refactoring the tests. Now that I use my own no-op operation, I can test for both direct and cyclical recursions.